File: predexonooor/engine.py
from __future__ import annotations

import logging
import time
import uuid
from dataclasses import dataclass
from typing import Any

from .config import AppConfig
from .liquidity import score_top_of_book
from .predexon import DataClient, TradeClient
from .quotes import TopOfBook, quote_limitless_market, quote_polymarket_token
from .risk import RiskLimits, can_open_trade, initial_state
from .storage.sqlite import (
    connect as db_connect,
    exposures,
    init_db,
    list_open_positions,
    open_position,
    total_exposure,
)


log = logging.getLogger("predexonooor")


@dataclass(frozen=True)
class VenueListing:
    venue: str
    token_id: str | None
    market_slug: str | None


class DataRateLimiter:
    def __init__(self, min_interval_seconds: float) -> None:
        self._min_interval = min_interval_seconds
        self._last = 0.0

    def wait(self) -> None:
        now = time.time()
        delta = now - self._last
        if delta < self._min_interval:
            time.sleep(self._min_interval - delta)
        self._last = time.time()


def _parse_listings(outcome: dict[str, Any]) -> list[VenueListing]:
    venues = outcome.get("venues") or []
    out: list[VenueListing] = []
    for v in venues:
        out.append(
            VenueListing(
                venue=str(v.get("venue")),
                token_id=str(v.get("token_id")) if v.get("token_id") else None,
                market_slug=str(v.get("market_slug")) if v.get("market_slug") else None,
            )
        )
    return out


def _quote_listing(data: DataClient, listing: VenueListing) -> TopOfBook | None:
    if listing.venue == "polymarket" and listing.token_id:
        return quote_polymarket_token(data, token_id=listing.token_id)
    if listing.venue == "limitless" and listing.market_slug:
        return quote_limitless_market(data, market_slug=listing.market_slug)
    return None


def _pick_best_liquidity(
    data: DataClient, *, predexon_id: str, enabled_venues: set[str], limiter: DataRateLimiter
) -> tuple[TopOfBook, VenueListing] | None:
    limiter.wait()
    outcome = data.get_outcome(predexon_id=predexon_id, routable_only=True)
    listings = [l for l in _parse_listings(outcome) if l.venue in enabled_venues]
    best: tuple[float, TopOfBook, VenueListing] | None = None

    for listing in listings:
        limiter.wait()
        q = _quote_listing(data, listing)
        if not q:
            continue
        s = score_top_of_book(q)
        if best is None or s.score > best[0]:
            best = (s.score, q, listing)

    if not best:
        return None
    return best[1], best[2]


def _extract_market_candidates(markets_payload: dict[str, Any]) -> list[dict[str, Any]]:
    markets = markets_payload.get("markets") or []
    out: list[dict[str, Any]] = []
    for m in markets:
        outcomes = m.get("outcomes") or []
        for o in outcomes:
            predexon_id = o.get("predexon_id") or o.get("predexonId")
            token_id = o.get("token_id") or o.get("tokenId")
            price = o.get("price")
            if not predexon_id or not token_id or price is None:
                continue
            p = float(price)
            if p <= 0.05 or p >= 0.95:
                continue
            out.append(
                {
                    "title": m.get("title") or m.get("question") or "",
                    "predexon_id": str(predexon_id),
                    "token_id": str(token_id),
                }
            )
    return out


def run_bot(cfg: AppConfig, *, data: DataClient, trade: TradeClient, db_path: str = "state.db") -> None:
    conn = db_connect(db_path)
    init_db(conn)

    enabled_venues = set(cfg.venues.enabled)
    limiter = DataRateLimiter(min_interval_seconds=1.1)

    limits = RiskLimits(
        starting_capital_usd=cfg.risk.starting_capital_usd,
        max_per_trade_usd=cfg.risk.max_per_trade_usd,
        max_exposure_per_market_usd=cfg.risk.max_exposure_per_market_usd,
        max_total_exposure_usd=cfg.risk.max_total_exposure_usd,
        max_open_positions=cfg.risk.max_open_positions,
        daily_max_loss_usd=cfg.risk.daily_max_loss_usd,
        max_drawdown_usd=cfg.risk.max_drawdown_usd,
    )
    risk_state = initial_state(limits)

    if cfg.mode not in {"dry_run", "live"}:
        raise ValueError("mode must be dry_run or live")
    if cfg.mode == "live" and not cfg.account_id:
        raise ValueError("account_id is required in live mode")

    log.info("boot health data=%s trade=%s", data.health(), trade.health())

    while True:
        open_pos = list_open_positions(conn)
        exp_by_market = exposures(conn)
        exp_total = total_exposure(conn)

        for pos in open_pos:
            q: TopOfBook | None = None
            if pos.venue == "polymarket":
                limiter.wait()
                q = quote_polymarket_token(data, token_id=pos.token_id)
            elif pos.venue == "limitless":
                limiter.wait()
                outcome = data.get_outcome(predexon_id=pos.predexon_id, routable_only=True)
                listing = next((l for l in _parse_listings(outcome) if l.venue == "limitless"), None)
                if listing and listing.market_slug:
                    limiter.wait()
                    q = quote_limitless_market(data, market_slug=listing.market_slug)
            if not q:
                continue

            now = int(time.time())
            should_exit = False
            reason = ""
            if q.mid >= pos.take_profit:
                should_exit = True
                reason = "take_profit"
            elif q.mid <= pos.stop_loss:
                should_exit = True
                reason = "stop_loss"
            elif now >= pos.max_hold_until:
                should_exit = True
                reason = "max_hold"

            if not should_exit:
                continue

            exit_price = max(0.001, min(0.999, q.best_bid * (1 - cfg.execution.slippage_guard_pct)))
            log.info(
                "exit signal pos=%s reason=%s venue=%s mid=%.3f bid=%.3f ask=%.3f",
                pos.id,
                reason,
                pos.venue,
                q.mid,
                q.best_bid,
                q.best_ask,
            )
            if cfg.mode == "live":
                client_id = str(uuid.uuid4())
                trade.place_order(
                    account_id=cfg.account_id,
                    venue=pos.venue,  # type: ignore[arg-type]
                    token_id=pos.token_id,
                    side="sell",
                    order_type="limit",
                    size=pos.size,
                    price=exit_price,
                    client_id=client_id,
                )
            conn.execute("UPDATE positions SET status='closed' WHERE id=?", (pos.id,))
            conn.commit()

        open_pos = list_open_positions(conn)
        if len(open_pos) < cfg.risk.max_open_positions:
            limiter.wait()
            markets_payload = data.list_polymarket_markets(limit=cfg.max_markets_scan)
            base_candidates = _extract_market_candidates(markets_payload)

            for c in base_candidates:
                predexon_id = c["predexon_id"]
                if any(p.predexon_id == predexon_id and p.status == "open" for p in open_pos):
                    continue

                picked = _pick_best_liquidity(data, predexon_id=predexon_id, enabled_venues=enabled_venues, limiter=limiter)
                if not picked:
                    continue
                q, listing = picked

                if q.spread > cfg.liquidity.max_spread:
                    continue
                if min(q.bid_size, q.ask_size) * q.mid < cfg.liquidity.min_top_depth_usd:
                    continue

                intended_notional = min(cfg.risk.max_per_trade_usd, cfg.risk.max_total_exposure_usd - exp_total)
                market_exp = exp_by_market.get(predexon_id, 0.0)
                ok, why = can_open_trade(
                    limits=limits,
                    state=risk_state,
                    open_positions=len(open_pos),
                    total_exposure=exp_total,
                    market_exposure=market_exp,
                    intended_notional=intended_notional,
                )
                if not ok:
                    continue

                token_id = listing.token_id
                if not token_id:
                    continue
                entry_price = min(0.999, q.best_ask * (1 + cfg.execution.slippage_guard_pct))
                size = max(0.01, intended_notional / entry_price)

                take_profit = entry_price * (1 + cfg.execution.entry.take_profit_pct)
                stop_loss = entry_price * (1 - cfg.execution.entry.stop_loss_pct)
                max_hold_until = int(time.time()) + (cfg.execution.entry.max_hold_minutes * 60)

                log.info(
                    "enter candidate venue=%s spread=%.4f depth=%.2f title=%s",
                    listing.venue,
                    q.spread,
                    min(q.bid_size, q.ask_size),
                    c["title"],
                )

                if cfg.mode == "live":
                    client_id = str(uuid.uuid4())
                    trade.place_order(
                        account_id=cfg.account_id,
                        venue=listing.venue,  # type: ignore[arg-type]
                        token_id=token_id,
                        side="buy",
                        order_type="limit",
                        size=size,
                        price=entry_price,
                        client_id=client_id,
                    )

                open_position(
                    conn,
                    predexon_id=predexon_id,
                    venue=listing.venue,
                    token_id=token_id,
                    size=size,
                    entry_price=entry_price,
                    take_profit=take_profit,
                    stop_loss=stop_loss,
                    max_hold_until=max_hold_until,
                )
                open_pos = list_open_positions(conn)
                exp_by_market = exposures(conn)
                exp_total = total_exposure(conn)
                if len(open_pos) >= cfg.risk.max_open_positions:
                    break

        time.sleep(cfg.poll_interval_seconds)

File: src/engine.ts
import crypto from "node:crypto";

import { AppConfig } from "./config.js";
import { RateLimiter } from "./limiter.js";
import { scoreTopOfBook } from "./liquidity.js";
import { quoteLimitlessMarket, quotePolymarketToken, mid } from "./quotes.js";
import { DataClient, PredexonApiError, TradeClient } from "./predexon.js";
import { applyFillToLedger, exposuresByMarket, loadState, realizedToday, saveState, totalExposure, type BotState, type Fill } from "./state.js";
import { RiskLimits, canOpenTrade, initialRiskState } from "./risk.js";

type VenueListing = {
  venue: string;
  tokenId?: string;
  marketSlug?: string;
  side?: string;
};

const todayISO = (): string => {
  const d = new Date();
  const y = d.getFullYear();
  const m = String(d.getMonth() + 1).padStart(2, "0");
  const day = String(d.getDate()).padStart(2, "0");
  return `${y}-${m}-${day}`;
};

const toNumber = (v: any): number => {
  const n = Number(v);
  return Number.isFinite(n) ? n : 0;
};

const parseExecutedFill = (resp: any): { ok: boolean; filledSize: number; avgPrice: number; orderId: string } => {
  const status = String(resp?.status ?? "");
  const orderId = String(resp?.orderId ?? "");
  const filled = toNumber(resp?.filled);
  const price = toNumber(resp?.price);

  if (!orderId) return { ok: false, filledSize: 0, avgPrice: 0, orderId: "" };
  if (status !== "filled" && status !== "partial" && status !== "cancelled") return { ok: false, filledSize: 0, avgPrice: 0, orderId };
  if (filled <= 0 || price <= 0) return { ok: false, filledSize: 0, avgPrice: 0, orderId };
  return { ok: true, filledSize: filled, avgPrice: price, orderId };
};

const parseListings = (outcome: any): VenueListing[] => {
  const venues = (outcome?.venues ?? []) as any[];
  return venues.map((v) => ({
    venue: String(v?.venue ?? ""),
    tokenId: v?.token_id ? String(v.token_id) : undefined,
    marketSlug: v?.market_slug ? String(v.market_slug) : undefined,
    side: v?.side ? String(v.side) : undefined
  }));
};

const quoteListing = async (data: DataClient, listing: VenueListing) => {
  if (listing.venue === "polymarket" && listing.tokenId) return await quotePolymarketToken(data, listing.tokenId);
  if (listing.venue === "limitless" && listing.marketSlug) {
    const side = listing.side === "no" ? "no" : "yes";
    return await quoteLimitlessMarket(data, listing.marketSlug, side);
  }
  return null;
};

const pickBestLiquidity = async (opts: {
  data: DataClient;
  predexonId: string;
  enabledVenues: Set<string>;
  limiter: RateLimiter;
  onOutcome404?: (predexonId: string) => void;
}): Promise<
  | {
      venue: "polymarket" | "limitless";
      tokenId?: string;
      marketSlug?: string;
      quoteMid: number;
      bestAsk: number;
      bestBid: number;
      bidSize: number;
      askSize: number;
    }
  | null
> => {
  await opts.limiter.wait();
  let outcome: any;
  try {
    outcome = await opts.data.getOutcome(opts.predexonId, true);
  } catch (e: any) {
    if (e instanceof PredexonApiError && e.statusCode === 404) {
      if (opts.onOutcome404) opts.onOutcome404(opts.predexonId);
      return null;
    }
    process.stdout.write(
      JSON.stringify(
        { event: "outcome_error", predexonId: opts.predexonId, statusCode: e?.statusCode, message: String(e?.message ?? e) },
        null,
        2
      ) + "\n"
    );
    return null;
  }
  const listings = parseListings(outcome).filter((l) => opts.enabledVenues.has(l.venue));

  let best:
    | {
        score: number;
        venue: "polymarket" | "limitless";
        tokenId?: string;
        marketSlug?: string;
        quoteMid: number;
        bestAsk: number;
        bestBid: number;
        bidSize: number;
        askSize: number;
      }
    | null = null;

  for (const l of listings) {
    await opts.limiter.wait();
    let q: any = null;
    try {
      q = await quoteListing(opts.data, l);
    } catch (e: any) {
      process.stdout.write(
        JSON.stringify(
          { event: "quote_error", predexonId: opts.predexonId, venue: l.venue, statusCode: e?.statusCode, message: String(e?.message ?? e) },
          null,
          2
        ) + "\n"
      );
      continue;
    }
    if (!q) continue;
    const s = scoreTopOfBook(q);
    const tokenId = q.venue === "polymarket" ? l.tokenId : undefined;
    const marketSlug = q.venue === "limitless" ? l.marketSlug : undefined;
    if (q.venue === "polymarket" && !tokenId) continue;
    if (q.venue === "limitless" && !marketSlug) continue;
    const m = mid(q);
    if (!best || s.score > best.score) {
      best = {
        score: s.score,
        venue: q.venue,
        tokenId,
        marketSlug,
        quoteMid: m,
        bestAsk: q.bestAsk,
        bestBid: q.bestBid,
        bidSize: q.bidSize,
        askSize: q.askSize
      };
    }
  }

  return best;
};

const extractMarketCandidates = (
  marketsPayload: any,
  filter: { minOutcomePrice: number; maxOutcomePrice: number }
): Array<{ title: string; predexonId: string; tokenId?: string }> => {
  const markets = (marketsPayload?.markets ?? []) as any[];
  const out: Array<{ title: string; predexonId: string; tokenId?: string }> = [];
  for (const m of markets) {
    const outcomes = (m?.outcomes ?? []) as any[];
    for (const o of outcomes) {
      const predexonId = o?.predexon_id ?? o?.predexonId;
      const tokenId = o?.token_id ?? o?.tokenId;
      const price = o?.price;
      if (!predexonId || price === undefined || price === null) continue;
      const p = Number(price);
      if (!Number.isFinite(p)) continue;
      if (p <= filter.minOutcomePrice || p >= filter.maxOutcomePrice) continue;
      out.push({
        title: String(m?.title ?? m?.question ?? ""),
        predexonId: String(predexonId),
        tokenId: tokenId ? String(tokenId) : undefined
      });
    }
  }
  return out;
};

export const runBot = async (cfg: AppConfig, opts: { data: DataClient; trade: TradeClient; statePath: string }) => {
  const enabledVenues = new Set(cfg.venues.enabled);
  const polymarketOnly = enabledVenues.size === 1 && enabledVenues.has("polymarket");
  const limiter = new RateLimiter(cfg.requestIntervalMs);

  const limits: RiskLimits = { ...cfg.risk };
  const riskState = initialRiskState(limits);

  const state: BotState = loadState(opts.statePath);

  const sleep = (ms: number) => new Promise((r) => setTimeout(r, ms));
  const isAbortError = (e: any) => String(e?.name ?? "") === "AbortError";

  saveState(opts.statePath, state);

  try {
    const health = { data: await opts.data.health(), trade: await opts.trade.health() };
    process.stdout.write(JSON.stringify({ health }, null, 2) + "\n");
  } catch (e: any) {
    process.stdout.write(
      JSON.stringify({ event: "health_error", name: String(e?.name ?? ""), message: String(e?.message ?? e) }, null, 2) + "\n"
    );
  }

  const outcome404Until = new Map<string, number>();
  const outcome404TtlMs = Math.max(0, cfg.outcome404TtlMinutes) * 60_000;
  let outcome404Count = 0;
  let lastOutcome404LogAt = 0;

  const onOutcome404 = (predexonId: string): void => {
    outcome404Until.set(predexonId, Date.now() + outcome404TtlMs);
    outcome404Count += 1;
    const now = Date.now();
    if (now - lastOutcome404LogAt >= 60_000) {
      lastOutcome404LogAt = now;
      process.stdout.write(JSON.stringify({ event: "skip_outcome_404_batch", count: outcome404Count }, null, 2) + "\n");
      outcome404Count = 0;
    }
  };

  const syncRealizedRisk = (): void => {
    const day = todayISO();
    if (riskState.dayKey !== day) riskState.dayKey = day;
    riskState.realizedPnlToday = realizedToday(state, day);
  };

  while (true) {
    syncRealizedRisk();
    const openPositions = state.positions.filter((p) => p.status === "open");
    const tickStartedAt = Date.now();
    const tickStats: Record<string, any> = {
      openPositions: openPositions.length,
      candidates: 0,
      candidatesMissingTokenId: 0,
      skippedOutcome404: 0,
      bestNull: 0,
      rejectSpread: 0,
      rejectDepth: 0,
      rejectRisk: 0,
      entered: 0
    };

    for (const pos of openPositions) {
      let qMid: number | null = null;
      let bestBid: number | null = null;
      let bestAsk: number | null = null;

      await limiter.wait();
      if (pos.venue === "polymarket") {
        if (!pos.tokenId) continue;
        let q: any = null;
        try {
          q = await quotePolymarketToken(opts.data, pos.tokenId);
        } catch (e: any) {
          process.stdout.write(
            JSON.stringify(
              { event: "quote_error", predexonId: pos.predexonId, venue: "polymarket", statusCode: e?.statusCode, message: String(e?.message ?? e) },
              null,
              2
            ) + "\n"
          );
          continue;
        }
        if (q) {
          qMid = mid(q);
          bestBid = q.bestBid;
          bestAsk = q.bestAsk;
        }
      } else if (pos.venue === "limitless") {
        let outcome: any;
        try {
          outcome = await opts.data.getOutcome(pos.predexonId, true);
        } catch (e: any) {
          if (e instanceof PredexonApiError && e.statusCode === 404) continue;
          process.stdout.write(
            JSON.stringify(
              { event: "outcome_error", predexonId: pos.predexonId, statusCode: e?.statusCode, message: String(e?.message ?? e) },
              null,
              2
            ) + "\n"
          );
          continue;
        }
        const listing = parseListings(outcome).find((l) => l.venue === "limitless");
        if (listing?.marketSlug) {
          const side = listing.side === "no" ? "no" : "yes";
          let q: any = null;
          try {
            q = await quoteLimitlessMarket(opts.data, listing.marketSlug, side);
          } catch (e: any) {
            process.stdout.write(
              JSON.stringify(
                { event: "quote_error", predexonId: pos.predexonId, venue: "limitless", statusCode: e?.statusCode, message: String(e?.message ?? e) },
                null,
                2
              ) + "\n"
            );
            continue;
          }
          if (q) {
            qMid = mid(q);
            bestBid = q.bestBid;
            bestAsk = q.bestAsk;
          }
        }
      }

      if (qMid === null || bestBid === null || bestAsk === null) continue;

      const now = Date.now();
      const shouldExit =
        qMid >= pos.takeProfit || qMid <= pos.stopLoss || now >= pos.maxHoldUntil;

      if (!shouldExit) continue;

      const exitPrice = Math.max(0.001, Math.min(0.999, bestBid * (1 - cfg.execution.slippageGuardPct)));
      process.stdout.write(
        JSON.stringify(
          {
            event: "exit_signal",
            posId: pos.id,
            venue: pos.venue,
            mid: qMid,
            bestBid,
            bestAsk,
            exitPrice
          },
          null,
          2
        ) + "\n"
      );

      if (cfg.mode === "live") {
        const clientId = crypto.randomUUID();
        const resp = await opts.trade.placeOrder({
          accountId: cfg.accountId,
          venue: pos.venue,
          predexonId: pos.venue === "polymarket" ? undefined : pos.predexonId,
          market: pos.venue === "polymarket" ? { tokenId: pos.tokenId } : undefined,
          side: "sell",
          type: "market",
          size: pos.size,
          clientId
        });
        const exec = parseExecutedFill(resp);
        if (exec.ok) {
          const fill: Fill = {
            id: crypto.randomUUID(),
            ts: Date.now(),
            dayISO: todayISO(),
            accountId: cfg.accountId,
            venue: pos.venue,
            predexonId: pos.predexonId,
            side: "sell",
            filledSize: exec.filledSize,
            avgPrice: exec.avgPrice,
            orderId: exec.orderId,
            clientId
          };
          applyFillToLedger(state, fill);
          syncRealizedRisk();
          if (exec.filledSize + 1e-9 < pos.size) {
            pos.size -= exec.filledSize;
            saveState(opts.statePath, state);
            continue;
          }
        }
      } else {
        const fill: Fill = {
          id: crypto.randomUUID(),
          ts: Date.now(),
          dayISO: todayISO(),
          accountId: cfg.accountId || "dry_run",
          venue: pos.venue,
          predexonId: pos.predexonId,
          side: "sell",
          filledSize: pos.size,
          avgPrice: exitPrice,
          orderId: `dryrun:${crypto.randomUUID()}`,
          clientId: "dryrun"
        };
        applyFillToLedger(state, fill);
        syncRealizedRisk();
      }

      pos.status = "closed";
      saveState(opts.statePath, state);
    }

    let openAfter = state.positions.filter((p) => p.status === "open");
    if (openAfter.length < cfg.risk.maxOpenPositions) {
      let marketsPayload: any;
      try {
        await limiter.wait();
        marketsPayload = await opts.data.listDiscoveryMarkets({ limit: cfg.maxMarketsScan });
      } catch (e: any) {
        process.stdout.write(
          JSON.stringify(
            { event: "discovery_error", name: String(e?.name ?? ""), message: String(e?.message ?? e), retryable: isAbortError(e) },
            null,
            2
          ) + "\n"
        );
        await sleep(isAbortError(e) ? 5_000 : 2_000);
        continue;
      }
      const candidates = extractMarketCandidates(marketsPayload, cfg.candidate);
      tickStats.candidates = candidates.length;

      for (const c of candidates) {
        openAfter = state.positions.filter((p) => p.status === "open");
        if (openAfter.length >= cfg.risk.maxOpenPositions) break;
        if (openAfter.some((p) => p.predexonId === c.predexonId && p.status === "open")) continue;

        const expByMarket = exposuresByMarket(state);
        const expTotal = totalExposure(state);

        if (!polymarketOnly) {
          const blockedUntil = outcome404Until.get(c.predexonId);
          if (blockedUntil && blockedUntil > Date.now()) {
            tickStats.skippedOutcome404 += 1;
            continue;
          }
        }

        let best:
          | {
              venue: "polymarket" | "limitless";
              tokenId?: string;
              marketSlug?: string;
              quoteMid: number;
              bestAsk: number;
              bestBid: number;
              bidSize: number;
              askSize: number;
            }
          | null = null;

        if (polymarketOnly) {
          if (!c.tokenId) {
            tickStats.candidatesMissingTokenId += 1;
            continue;
          }
          await limiter.wait();
          let q: any = null;
          try {
            q = await quotePolymarketToken(opts.data, c.tokenId);
          } catch (e: any) {
            process.stdout.write(
              JSON.stringify(
                { event: "quote_error", predexonId: c.predexonId, venue: "polymarket", statusCode: e?.statusCode, message: String(e?.message ?? e) },
                null,
                2
              ) + "\n"
            );
            continue;
          }
          if (!q) continue;
          best = {
            venue: "polymarket",
            tokenId: c.tokenId,
            quoteMid: mid(q),
            bestAsk: q.bestAsk,
            bestBid: q.bestBid,
            bidSize: q.bidSize,
            askSize: q.askSize
          };
        } else {
          best = await pickBestLiquidity({
            data: opts.data,
            predexonId: c.predexonId,
            enabledVenues,
            limiter,
            onOutcome404
          });
        }
        if (!best) {
          tickStats.bestNull += 1;
          continue;
        }

        const spread = best.bestAsk - best.bestBid;
        if (spread > cfg.liquidity.maxSpread) {
          tickStats.rejectSpread += 1;
          continue;
        }

        const depthShares = Math.min(best.bidSize, best.askSize);
        if (depthShares * best.quoteMid < cfg.liquidity.minTopDepthUsd) {
          tickStats.rejectDepth += 1;
          continue;
        }

        const intendedNotional = Math.min(cfg.risk.maxPerTradeUsd, cfg.risk.maxTotalExposureUsd - expTotal);
        if (intendedNotional <= 0) continue;
        const marketExp = expByMarket[c.predexonId] ?? 0;
        const gate = canOpenTrade({
          limits,
          state: riskState,
          openPositions: openAfter.length,
          totalExposure: expTotal,
          marketExposure: marketExp,
          intendedNotional
        });
        if (!gate.ok) {
          tickStats.rejectRisk += 1;
          continue;
        }

        process.stdout.write(
          JSON.stringify(
            {
              event: "enter_signal",
              venue: best.venue,
              title: c.title,
              predexonId: c.predexonId,
              spread,
              bestBid: best.bestBid,
              bestAsk: best.bestAsk,
              intendedNotional
            },
            null,
            2
          ) + "\n"
        );
        tickStats.entered += 1;

        if (cfg.mode === "live") {
          const clientId = crypto.randomUUID();
          const resp = await opts.trade.placeOrder({
            accountId: cfg.accountId,
            venue: best.venue,
            predexonId: best.venue === "polymarket" ? undefined : c.predexonId,
            market: best.venue === "polymarket" ? { tokenId: best.tokenId } : undefined,
            side: "buy",
            type: "market",
            amount: intendedNotional,
            clientId
          });
          const exec = parseExecutedFill(resp);
          if (!exec.ok) {
            continue;
          }
          const entryPrice = exec.avgPrice;
          const filledSize = exec.filledSize;
          const tp = entryPrice * (1 + cfg.execution.entry.takeProfitPct);
          const sl = entryPrice * (1 - cfg.execution.entry.stopLossPct);
          const maxHoldUntil = Date.now() + cfg.execution.entry.maxHoldMinutes * 60_000;

          const fill: Fill = {
            id: crypto.randomUUID(),
            ts: Date.now(),
            dayISO: todayISO(),
            accountId: cfg.accountId,
            venue: best.venue,
            predexonId: c.predexonId,
            side: "buy",
            filledSize,
            avgPrice: entryPrice,
            orderId: exec.orderId,
            clientId
          };
          applyFillToLedger(state, fill);
          syncRealizedRisk();

          state.positions.push({
            id: crypto.randomUUID(),
            predexonId: c.predexonId,
            venue: best.venue,
            tokenId: best.venue === "polymarket" ? best.tokenId : undefined,
            size: filledSize,
            entryPrice,
            entryTs: Date.now(),
            takeProfit: tp,
            stopLoss: sl,
            maxHoldUntil,
            status: "open"
          });
          saveState(opts.statePath, state);
          continue;
        }

        const entryPrice = Math.min(0.999, best.bestAsk * (1 + cfg.execution.slippageGuardPct));
        const size = Math.max(0.01, intendedNotional / entryPrice);
        const tp = entryPrice * (1 + cfg.execution.entry.takeProfitPct);
        const sl = entryPrice * (1 - cfg.execution.entry.stopLossPct);
        const maxHoldUntil = Date.now() + cfg.execution.entry.maxHoldMinutes * 60_000;

        const fill: Fill = {
          id: crypto.randomUUID(),
          ts: Date.now(),
          dayISO: todayISO(),
          accountId: cfg.accountId || "dry_run",
          venue: best.venue,
          predexonId: c.predexonId,
          side: "buy",
          filledSize: size,
          avgPrice: entryPrice,
          orderId: `dryrun:${crypto.randomUUID()}`,
          clientId: "dryrun"
        };
        applyFillToLedger(state, fill);
        syncRealizedRisk();

        state.positions.push({
          id: crypto.randomUUID(),
          predexonId: c.predexonId,
          venue: best.venue,
          tokenId: best.venue === "polymarket" ? best.tokenId : undefined,
          size,
          entryPrice,
          entryTs: Date.now(),
          takeProfit: tp,
          stopLoss: sl,
          maxHoldUntil,
          status: "open"
        });
        saveState(opts.statePath, state);
      }
    }

    process.stdout.write(
      JSON.stringify(
        {
          event: "tick_summary",
          polymarketOnly,
          pollIntervalSeconds: cfg.pollIntervalSeconds,
          maxMarketsScan: cfg.maxMarketsScan,
          requestIntervalMs: cfg.requestIntervalMs,
          minOutcomePrice: cfg.candidate.minOutcomePrice,
          maxOutcomePrice: cfg.candidate.maxOutcomePrice,
          maxSpread: cfg.liquidity.maxSpread,
          minTopDepthUsd: cfg.liquidity.minTopDepthUsd,
          ...tickStats,
          tickMs: Date.now() - tickStartedAt
        },
        null,
        2
      ) + "\n"
    );

    await new Promise((r) => setTimeout(r, cfg.pollIntervalSeconds * 1000));
  }
};
